text/0000-atcompoundkey-error-path-optimization.md

```diff
@@ -0,0 +1,199 @@
+
+The following graph shows the difference between the performance of happy path and error path in AtCompoundKey API
+
+<div>
+    <a href="https://plot.ly/~karthikreddy09/1/?share_key=siExwBF3DybX1HhrW4dHe0" target="_blank" title="Plot 1" style="display: block; text-align: center;"><img src="https://plot.ly/~karthikreddy09/1.png?share_key=siExwBF3DybX1HhrW4dHe0" alt="Plot 1" style="max-width: 100%;width: 600px;"  width="600" onerror="this.onerror=null;this.src='https://plot.ly/404.png';" /></a>
+    <script data-plotly="karthikreddy09:1" sharekey-plotly="siExwBF3DybX1HhrW4dHe0" src="https://plot.ly/embed.js" async></script>
+</div>
+
+
+## Option 1 - Modify existing API to return a empty const Any& if the key is not found
+Remove the non-const version of the API and retain just the const version. Return empty `Any` by reference if the key is not found
+
+<strike>Any& AtCompoundkey(const std::string& key);</strike>
+
+`const Any& AtCompoundKey(const std::string& key) const;`
+
+### Pros
+* Addresses the performance problem. The happy path and the error path have similar performance. 
+
+### Cons
+* Happy path has a slight performance decrease due to the use of 'find' API instead of the 'at' API
+* Although the API returns a const reference, the user could add back a non-const version by const_cast'ing the returned ref, which will break every subsequent call. Infact the current non-const API does exactly this, which is ikky.
+* The information regarding the error conditions is lost. However, do the users really care about the information is a puzzle.
+* If the BundleManifest class is modified to handle "null" values in the manifest, this API has to change.  
+
+## Option 2 - Modify the current API to return an Any object by value
+Return an empty `Any` object if the key is not found
+
+```
+Any AtCompoundKey(const std::string& key);
+```
+
+### Pros
+* Addresses the performance problem. The happy path and the error path have similar performance. 
+
+### Cons
+* Returning by value results in copies of the returned object which could be costly, specially if this API is used to retrieve intermediate maps of larger size.
+* It is not clear how a null value from a manifest file is treated. specifying "null" is legal in JSON. Current implementation of bundle metadata parser ignores the entries with null value. However, we cannot assume this will not change in the future.
+* The information regarding the error conditions is lost.
+
+## Option 2.1 (Preferred) - Add an overload with a default value parameter 
```

Comment:
If we finally decide against the JSON API approach, I think a combination of 1. and 2.1. would make sense. Basically like the `std::optional` API and providing a `const& Any AtCompoundKey(const std::string&) const` as well as a `Any AtCompoundKeyOr(const std::string&, Any) const` function.